src/shared/quota-util.c

```diff
@@ -34,3 +35,49 @@ int quotactl_fd_with_fallback(int fd, int cmd, int id, void *addr) {
 
         return RET_NERRNO(quotactl(cmd, devnode, id, addr));
 }
+
+int quota_query_proj_id(int fd, uint32_t proj_id, struct dqblk *req) {
+        int r;
+
+        assert(fd >= 0);
+        assert(req);
+
+        zero(*req);
+
+        r = RET_NERRNO(quotactl_fd(fd, QCMD_FIXED(Q_GETQUOTA, PRJQUOTA), proj_id, req));
+        if (r == -ESRCH || ERRNO_IS_NEG_NOT_SUPPORTED(r) || ERRNO_IS_NEG_PRIVILEGE(r))
+                return false;
+        if (r < 0)
+                return r;
+
+        return true;
+}
+
+int quota_proj_id_set_recursive(int fd, uint32_t proj_id, bool verify_exclusive) {
+        int r;
+
+        assert(fd >= 0);
+
+        /* Set to top level first because of the case where directories already exist with multiple subdirectories,
+           in which case, number of inodes will be > 1 if applied recursively only */
+        r = set_proj_id(fd, proj_id);
+        if (r < 0)
+                return r;
+
+        /* Confirm only the current inode has the project id (in case of race conditions) */
+        if (verify_exclusive) {
```

Comment:
hmm, the set_proj_id() above should be within this if block, no? not before?

Reply:
but the set_proj_id_recursive() below should be a superset of what set_proj_id() does, no? hence the set_proj_id() is redundant, if verify_exclusive is false?